src/ocrd_network/cli/client.py

```diff
@@ -1,11 +1,43 @@
 import click
+from http.server import BaseHTTPRequestHandler, HTTPServer
 from typing import Optional
 
 from ocrd.decorators import parameter_option
 from ocrd_network import Client
 from ocrd_utils import DEFAULT_METS_BASENAME
 
 
+STOP_WAITING_CALLBACK = False
+
+
+class ClientCallbackHandler(BaseHTTPRequestHandler):
+    """
+    A simple callback handler for the network client to be invoked when the Processing Worker
+    sends requests to the `callback_url` set in the processing request submitted to the Processing Server.
+    """
+
+    def do_POST(self):
+        self.send_response(200)
+        self.send_header("Content-Type", "text/plain")
+        self.end_headers()
+        self.wfile.write("finished".encode("utf-8"))
+        len = int(self.headers.get("Content-Length", 0))
+        data = self.rfile.read(len).decode("utf-8")
+        # TODO: how should the callback-content be handled/printed
+        print(f"Processor finished: {data}")
+        global STOP_WAITING_CALLBACK
+        STOP_WAITING_CALLBACK = True
+
+
+class ClientCallbackServer(HTTPServer):
+    """
+    A simple http-server that listens for callbacks from the Processing Server/Worker.
+    """
+    def __init__(self):
+        super().__init__(server_address=("0.0.0.0", 0), RequestHandlerClass=ClientCallbackHandler)
+        self.callback_url = f"http://172.17.0.1:{self.server_address[1]}"
```

Comment:
What is the purpose of `self.callback_url` of the callback server? And why does it use the docker host IP but with the same port as the callback server?

Is it so that the (remote) processing server can "find" the local callback server on the network - which would not work with `0.0.0.0`/`localhost`?

In any case we should make this configurable or deduce it from the environment, otherwise it will not work if Docker is set up differently.

Also, I'm wondering how well this scales, the limiting factor is the number of free ports?

Reply:
> What is the purpose of self.callback_url of the callback server? 

It is for setting the `user_callback` field in the processing request, check [here](https://github.com/OCR-D/core/blob/1c8b9db05431d9a52e4763b9efcf4f123e3d65c8/src/ocrd_network/processing_worker.py#L237) and [here](https://github.com/OCR-D/core/blob/master/src/ocrd_network/models/job.py#L20)

> And why does it use the docker host IP but with the same port as the callback server? Is it so that the (remote) processing server can "find" the local callback server on the network - which would not work with `0.0.0.0`/`localhost`?

Using the localhost also fails by reaching timeout. I tried with a docker host IP although I knew it would probably fail here. It does. It works with the setup of @joschrew because 1) he runs the client in a native environment and 2) the processing worker runs in a docker environment without defining a separate network (not exactly sure for 2)).

Things are different when the client runs in a docker environment as in the integration test. The default docker host is no longer working since there is a defined separate network in the docker-compose file named `ocrd_network_test` in the `.env`. That network's host address is different than the default docker host address. The host for the processing server is defined as a string in the docker-compose file. The address resolution is then handled correctly by docker. The processing server URL at run-time becomes: `http://ps-docker-host.ocrd_network_test:8000`. Similarly, I assume that it is needed to define a string host for the client (`http://client-docker-host.ocrd_network_test:0`) so the processing workers can reach it within the same docker network.

> Also, I'm wondering how well this scales, the limiting factor is the number of free ports?

That should be the case, yes. But since we have a separate docker network there should still be more free ports than the localhost.

Reply:
It will be still more complicated to handle the cases when the client runs natively since the callback address changes. The user should not be bothered by configuring that correctly. It should work by default. There is currently the `internal_callback_url` in the processing server config file. I think we would need one for the `client_callback_url` as well for switching between the docker and native environments. The polling still remains the much simpler solution.

Reply:
I have implemented the polling client. Processing and Workflow tests have passed now.

Reply:
Yeah, polling is much less error-prone, thanks for changing.